### PR TITLE
Optimize interpreter: Evaluate expressions directly without trampolines

### PR DESCRIPTION
**Pull request by turbolent**
⚠️ Depends on #680
Work towards #678

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -1538,269 +1497,184 @@ func (interpreter *Interpreter) assignmentGetterSetter(target ast.Expression) Tr
...
 }
 
-func (interpreter *Interpreter) visitBinaryOperation(
-	expr *ast.BinaryExpression,
-	f func(left, right Value) Value,
-) Trampoline {
-	left := interpreter.runAllStatements(expr.Left.Accept(interpreter).(Trampoline)).(Value)
-	right := interpreter.runAllStatements(expr.Right.Accept(interpreter).(Trampoline)).(Value)
-	return Done{Result: f(left, right)}
+func (interpreter *Interpreter) evalExpression(expression ast.Expression) Value {
```
Comment (SupunS) [inline runtime/interpreter/interpreter.go:1552]: oh this is already added. nice! :+1:

**Review by SupunS — APPROVED** [review]
LGTM!

File: runtime/interpreter/interpreter.go [hunk 2]
```diff
@@ -1840,59 +1714,49 @@ func (interpreter *Interpreter) testEqual(left, right Value) BoolValue {
...
-			return ExpressionStatementResult{value}
-		})
+	result := interpreter.evalExpression(statement.Expression)
+	var value Value
+	var ok bool
+	value, ok = result.(Value)
+	if !ok {
+		value = nil
+	}
+	return Done{Result: ExpressionStatementResult{value}}
```
Comment (SupunS) [inline runtime/interpreter/interpreter.go:1747]:
```suggestion
	return Done{
		Result: ExpressionStatementResult{result},
	}
```

File: runtime/interpreter/interpreter.go [hunk 3]
```diff
@@ -1372,30 +1343,28 @@ func (interpreter *Interpreter) VisitVariableDeclaration(declaration *ast.Variab
 	valueType := interpreter.Program.Elaboration.VariableDeclarationValueTypes[declaration]
 	secondValueType := interpreter.Program.Elaboration.VariableDeclarationSecondValueTypes[declaration]
 
-	return interpreter.visitPotentialStorageRemoval(declaration.Value).
-		FlatMap(func(result interface{}) Trampoline {
+	result := interpreter.visitPotentialStorageRemoval(declaration.Value)
 
-			valueCopy := interpreter.copyAndConvert(result.(Value), valueType, targetType)
+	valueCopy := interpreter.copyAndConvert(result.(Value), valueType, targetType)
```
Comment (SupunS) [inline runtime/interpreter/interpreter.go:1348]:
```suggestion
	valueCopy := interpreter.copyAndConvert(result, valueType, targetType)
```

File: runtime/interpreter/interpreter.go (hunk 2 above)
Comment (turbolent) [reply runtime/interpreter/interpreter.go:1747]: Good idea, nice!

File: runtime/interpreter/interpreter.go (hunk 3 above)
Comment (turbolent) [reply runtime/interpreter/interpreter.go:1348]: Good catch, this is not needed 👍

**Pull request merged**